pytorch_pfn_extras/utils/comparer.py

```diff
@@ -296,3 +300,197 @@ def _add_target(self, handle, models, outputs):
+                out1 = self._targets[backend1][val_name]
+                out2 = self._targets[backend2][val_name]
+                self._compare_fn(backend1, backend2, val_name, out1, out2)
+
+    def _compare_targets(self, handle, models, batch_idx, outputs):
+        engine, _, _ = self._engines[handle.name]
+        if hasattr(engine, "manager"):
+            class _ManagerProxy(manager_module._ManagerProxy):
+                @property
+                def iteration(self) -> int:
+                    return self._manager.iteration + 1
+
+            manager = _ManagerProxy(engine.manager)
+            if not self._trigger(manager):
+                return
+
+        # Save the outputs of this iteration
+        with self._report_lock:
+            self._add_target(handle, models, outputs)
+            if len(self._targets.keys()) == len(self._engines.keys()):
+                # all outputs have been filled, lets compare and reset
+                self._compare_outs()
+                self._targets = {}
+            self._assert_incompatible_trigger(not self._finalized)
+
+        # Excplicitly synchronize
+        self._semaphore.release()
+        self._barrier.wait()
+        self._semaphore.acquire()
+
+    def add_engine(self, name, engine, *args, **kwargs):
+        type_engine = type(engine)
+
+        if type_engine not in (_trainer.Trainer, _evaluator.Evaluator):
+            raise ValueError(f"Engine type {type_engine} is not supported")
+
+        if self._engine_type is None:
+            self._engine_type = type_engine
+        elif type_engine != self._engine_type:
+            raise ValueError("All the engines must be of the same type")
+
+        if name in self._engines.keys():
+            raise ValueError(f"Engine {engine} already registered")
+
+        self._engines[name] = engine, args, kwargs
+        engine.handler = _ComparableHandler(engine.handler, name, self._compare_targets)
+
+    def run_engine(self, engine, args, kwargs):
+        try:
+            self._semaphore.acquire()
+            engine.run(*args, **kwargs)
+            with self._report_lock:
+                self._finalized = True
+                self._assert_incompatible_trigger(len(self._targets) == 0)
+        except Exception:
+            self._barrier.abort()
+            raise
+        finally:
+            self._semaphore.release()
+
+    def dump(self, engine, dir, train_loader, val_loader):
+        raise NotImplementedError
+
+    def compare(self):
+        """Compares outputs.
+
+        Args:
+            loaders (dict of loaders):
+                Data loaders used as input for each engine.
+        """
+        n_workers = len(self._engines)
+        self._barrier = threading.Barrier(n_workers)
+        self._semaphore = threading.Semaphore(
+            n_workers if self._concurrency is None else self._concurrency)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=n_workers) as executor:
```

Comment:
Could you add a path to run without ThreadPoolExecutror when `self._concurrency == None`?